docs/articles/semantic_chunking.md

```diff
@@ -0,0 +1,537 @@
+from langchain import PromptTemplate
+from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline, BitsAndBytesConfig, AutoModel
+from langchain_community.llms import HuggingFaceHub
+from langchain.chains import create_extraction_chain_pydantic
+from pydantic import BaseModel
+from langchain_core.prompts import ChatPromptTemplate
+from typing import List
+from langchain.docstore.document import Document
+from langchain.output_parsers import PydanticOutputParser
+import json
+from pydantic import ValidationError
+import re
+import dspy
+import time
+
+class ExtractSentences(dspy.Signature):
+    """Extract meaningful propositions (semantic chunks) from the given text."""
+    text = dspy.InputField()
+    sentences = dspy.OutputField(desc="List of extracted sentences")
+
+class ExtractSentencesProgram(dspy.Program):
+    def run(self, text):
+        extract = dspy.Predict(ExtractSentences)
+        result = extract(text=text)
+        return result.sentences
+
+
+class LlmSemanticChunker(TextSplitter):
+    def __init__(self, llm, chunk_size: int = 1000):
+        super().__init__(chunk_size=chunk_size)
+        self.llm = llm
+        self.chunk_size = chunk_size  # Explicitly set chunk_size as an instance attribute
+        dspy.settings.configure(lm=llm)
+        self.extractor = ExtractSentencesProgram()
+
+    def get_propositions(self, text):
+        sentences = self.extractor.run(text)
+
+        if isinstance(sentences, list):
+            return sentences
+
+        # Fallback: extract sentences heuristically
+        return [s.strip() for s in text.split('.') if s.strip()]
+
+    def split_text(self, text: str) -> List[str]:
+        """Extract propositions and chunk them accordingly."""
+        propositions = self.get_propositions(text)
+        return self._chunk_propositions(propositions)
+
+    def split_documents(self, documents: List[Document]) -> List[Document]:
+        """Split documents into chunks."""
+        split_docs = []
+        for doc in documents:
+            chunks = self.split_text(doc.page_content)
+            for i, chunk in enumerate(chunks):
+                metadata = doc.metadata.copy()
+                metadata.update({"chunk_index": i})
+                split_docs.append(Document(page_content=chunk, metadata=metadata))
+        return split_docs
+
+    def _chunk_propositions(self, propositions: List[str]) -> List[str]:
+        chunks = []
+        current_chunk = []
+        current_size = 0
+
+        for prop in propositions:
+            prop_size = len(prop)
+            if current_size + prop_size > self.chunk_size and current_chunk:
+                chunks.append(" ".join(current_chunk))
+                current_chunk = []
+                current_size = 0
+            current_chunk.append(prop)
+            current_size += prop_size
+
+        if current_chunk:
+            chunks.append(" ".join(current_chunk))
+
+        return chunks
+```
+
+## Evaluation

```

Comment:
@Ashish-Abraham
doesn't it make sense to run the same text (as you did with semantic-embedding-based and hierarchical-clustering-based chunking) thru LLM-based chunking and show the output?

Reply:
Yes let me run it

Reply:
You can add this:
"""
Here is a sample output:
Chunk 1:
Regular exercise is essential for maintaining overall health and well-being It helps in controlling weight,
improving cardiovascular health, and boosting mental health Engaging in physical activity regularly can also enhance the immune system, reduce the risk of chronic diseases,
and increase energy levels Regular workouts are known to improve muscle strength and flexibility, which can prevent injuries and enhance mobility.
Moreover, exercise contributes to better sleep and improved mood, which are crucial for daily functioning
--------------------------------------------------
Chunk 2:
Physical activity can also help
reduce symptoms of anxiety and depression, leading to a more balanced emotional state Activities like walking, jogging, or swimming can be easily
incorporated into a daily routine, making it accessible for everyone By setting realistic goals and staying consistent, individuals can enjoy these benefits and lead a healthier lifestyle
--------------------------------------------------


Chunk 3:
Group fitness classes or sports teams can provide motivation and social support, making exercise more enjoyable and sustainable
--------------------------------------------------
'''